File: spider.py
# !/usr/bin/env python
# -*- coding:utf-8 -*-

import leancloud, auth
from datetime import datetime, timedelta
from leancloud import Object, Query, LeanCloudError
from lib import PyRSS2Gen
import recipe

Feed = Object.extend('FeedItem')
FeedInfo = Object.extend('FeedInfo')
OnlineFeed = Object.extend('OnlineFeed')
DebugLog = Object.extend('DebugLog')

def get_info(name):
    query = Query(FeedInfo).equal_to('name', name)
    try: info = query.first()
    except LeanCloudError, e:
        if e.code == 101:
            info = FeedInfo()
            info.set('name', name)
        else: raise(e)
    return info

def save_data(data):
    try: data.save()
    except LeanCloudError, e: print "Save feed error: %s" % str(e)

def set_feed_data(name, data):
    item = Feed()
    item.set('name', name)
    item.set('title', data[0])
    item.set('time', data[1])
    item.set('link', data[2])
    item.set('content', data[3])
    save_data(item)

def save_rss(name, rss, info):
    count = 0
    for data in rss.get_item():
        count += 1
        set_feed_data(name, data)
    print('Spider %s, add %s new feed' % (name, count))
    if count > 0: save_data(info)
    if len(rss.log) == 0: return
    log = DebugLog()
    log.set('name', name)
    for key,value in rss.log: log.set(str(key), value)
    save_data(log)

def save(all_feed=False, feed_name=None):
    for r in rss_list(all_feed):
        name = r.name
        if feed_name and feed_name != name: continue
        info = get_info(name)
        rss = r(info=info)
        try: save_rss(name, rss, info)
        except Exception as e: print('save %s fail : %s' % (r.name, str(e)))

def rss_list(all_feed=False):
    if all_feed: return (set(recipe.recipe_list) | set(recipe.hide_list))
    else: return set(recipe.recipe_list)

def save_online(feed_name=None):
    feeds = Query(OnlineFeed).find()
    base = recipe.base_list
    for r in feeds:
        base_class = base.get(r.get('base'), None)
        if not base_class: continue
        name = 'ol-%s' % r.get('name')
        if feed_name and feed_name != name: continue
        info = get_info(name)
        rss = base_class(info=info)
        rss.set_online_recipe(r.get('data'))
        try: save_rss(name, rss, info)
        except Exception as e: print('save %s fail : %s' % (r.name, str(e)))

def online_rss_list():
    return [a.get('name') for a in Query(OnlineFeed).find()]

def get_all_feed(name):
    query = Query(Feed)
    query.equal_to('name', name).descending("time")
    return query.find()

def show(name):
    rss = PyRSS2Gen.RSS2(title=name,link="https://github.com/miaowm5",description="RSSGen By Miaowm5")
    for e in get_all_feed(name):
        title = e.get('title')
        time = e.get('time')
        time = datetime(*(time.utctimetuple()[0:6]))
        link = e.get('link')
        content = e.get('content')
        item = PyRSS2Gen.RSSItem(title=title,pubDate=time,
          link=link,description=content)
        rss.items.append(item)
    return rss.to_xml(encoding='utf-8')

def clear():
    for r in rss_list(all_feed=True): clear_feed(r.name, r.oldest)

def clear_online():
    for r in Query(OnlineFeed).find(): clear_feed('ol-%s' % r.get('name'), r.get('data')[4])

def clear_feed(name, day):
    oldest = datetime.now() - timedelta(days=day)
    query = Query(Feed)
    query.equal_to('name', name).less_than("time", oldest)
    for e in query.find():
        print('delete old feed: %s (%s)' % (e.get('title').encode('utf-8'), e.get('time')))
        e.destroy()

def add_online_feed(name, base, data):
    try: feed = Query(OnlineFeed).equal_to('name', name).first()
    except LeanCloudError, e:
        if e.code == 101:
            feed = OnlineFeed()
            feed.set('name', name)
        else: raise(e)
    feed.set('base', base)
    feed.set('data', data)
    feed.save()

def get_online_feed(name):
    try: feed = Query(OnlineFeed).equal_to('name', name).first()
    except LeanCloudError, e:
        if e.code == 101: return [name, '', None]
        else: raise(e)
    return [feed.get('name'), feed.get('base'), feed.get('data')]

def remove_online_feed(name):
    Query(OnlineFeed).equal_to('name', name).first().destroy()
    for e in Query(Feed).equal_to('name', 'ol-%s' % name).find(): e.destroy()

File: spider.js

const {AV, Data} = require('./engine.js')

class Spider{
  async getList(){
    let appInfo = await (new AV.Query('AppInfo')).first()
    let list = appInfo.get('recipeList')
    return list
  }
  async getInfo(name){
    let self = this
    let query = (new AV.Query('FeedInfo')).equalTo('name', name)
    let info = null
    try{
      info = await query.first()
      if (!info) throw 'Info undefined'
    }catch(e){
      info = new Data['FeedInfo']()
      info.set('name', name)
      await self.save(info)
    }
    return info
  }
  async save(item){
    try{
      if (Array.isArray(item)){ await AV.Object.saveAll(item) }
      else{ await item.save() }
    }
    catch(e){
      console.error(`Meet error when save item`)
      console.error(e)
    }
  }
  async wait(time){ return new Promise((s,r)=>{ setTimeout(s, time) }) }
  async run(list=null){
    console.log('Spider start')
    let self = this
    if (list){ if (!Array.isArray(list)){ list = [list] } }
    else{ list = await this.getList() }
    console.log(`Recipe List:${list.join(', ')}` )
    list = list.map(async (name, index)=>{
      await self.wait(index*500)
      let info = await self.getInfo(name)
      let recipe = require('./recipe/'+name)
      recipe = new recipe(info)
      let feed
      try{ feed = await recipe.getFeed() }
      catch(e){
        console.error(`Get feed ${name} fail`)
        console.error(e)
        feed = []
      }
      if (feed.length > 0){
        feed = feed.map((feed)=>{
          let item = new Data['FeedItem']()
          item.set('name', name)
          item.set('title', feed.title)
          item.set('time', feed.date)
          item.set('link', feed.link)
          item.set('content', feed.content)
          return item
        })
        await self.save(feed)
        await self.save(info)
      }
      console.log(`Spider ${name}, find ${feed.length} new feed`)
      let log = recipe.log.map((log)=>{
        let item = new Data['Log']()
        item.set('name', name)
        item.set('type', log.type)
        item.set('info', log.info)
        return item
      })
      await self.save(log)
    })
    await Promise.all(list)
    console.log('Spider over')
  }
  async clear(){
    console.log('clear old data')
    let self = this
    const feed = async ()=>{
      let list = await this.getList()
      list = list.map(async (name, index)=>{
        await self.wait(index*500)
        let recipe = require('./recipe/'+name)
        recipe = new recipe()
        let oldest = new Date()
        oldest.setDate(oldest.getDate()-recipe.capture.oldest)
        let feed = (new AV.Query('FeedItem')).equalTo('name', name).lessThan("time", oldest)
        feed = await feed.find()
        await AV.Object.destroyAll(feed)
        if (feed.length > 0) console.log(`${recipe.name} delete ${feed.length} old feed`)
      })
      await Promise.all(list)
    }
    const log = async ()=>{
      let oldest = new Date()
      oldest.setDate(oldest.getDate()-3)
      let log = (new AV.Query('Log')).lessThan("time", oldest)
      log = await log.find()
      await AV.Object.destroyAll(log)
    }
    await Promise.all([feed(), log()])
    console.log('clear over')
  }
  async show(name, url){
    const Feed = require('feed')
    let feed = new Feed({
      title: name, description: name, link: url,
      updated: new Date(), generator: 'RSSGen'
    })
    let item = (new AV.Query('FeedItem')).equalTo('name', name).descending('time')
    item = await item.find()
    item.forEach((item)=>{
      let title = item.get('title')
      let link = item.get('link')
      let content = item.get('content')
      content += `<img src="${url}/qr?url=${encodeURIComponent(link)}&title=${encodeURIComponent(title)}"></img>`
      feed.addItem({
        title, link, content, description: content,
        date: item.get('time'),
      })
    })
    let result = feed.rss2()
    return result.replace('<rss version="2.0"',
      '<?xml-stylesheet type="text/xsl" href="../style/rss.xsl" ?><rss version="2.0"')
  }
}

module.exports = new Spider()
